File: HackerEarth/Sorting/merge-sort.cpp
#include <iostream>
#include <vector>

using namespace std;

long long merge(vector<long long>& arr, long long start, long long mid, long long end) {
    long long leftPosition = start;
    long long rightPosition = mid + 1;

    vector<long long> mergedArray;
    long long leftIsBiggerCnt = 0;

    while (leftPosition <= mid || rightPosition <= end) {
        if (leftPosition > mid)
            mergedArray.push_back(arr[rightPosition++]);
        else if (rightPosition > end)
            mergedArray.push_back(arr[leftPosition++]);
        else if (arr[leftPosition] <= arr[rightPosition])
            mergedArray.push_back(arr[leftPosition++]);
        else {
            mergedArray.push_back(arr[rightPosition++]);
            leftIsBiggerCnt += mid - leftPosition + 1;
        }
    }

    for (int i = start; i <= end; i++)
        arr[i] = mergedArray[i - start];

    return leftIsBiggerCnt;
}

long long split(vector<long long>& arr, long long start, long long end){
    if (start == end)
        return 0;

    long long leftIsBiggerCnt = 0;

    long long mid = (start + end) / 2;

    leftIsBiggerCnt += split(arr, start, mid);
    leftIsBiggerCnt += split(arr, mid + 1, end);

    leftIsBiggerCnt += merge(arr, start, mid, end);

    return leftIsBiggerCnt;
}

long long mergeSort(vector<long long>& arr) {
    return split(arr, 0, arr.size() - 1);
}

int main(){
    long long arraySize;

    cin >> arraySize;

    vector<long long> arr;

    for (long long i = 0; i < arraySize; i++){
        long long val;
        cin >> val;

        arr.push_back(val);
    }

    long long cnt = mergeSort(arr);
    cout << cnt;
}

File: HackerEarth/Sorting/merge-sort.py
def readIntArray():
    res = []
    for val in input().split():
        res.append(int(val))
    return res

def readInt():
    return int(input())

def merge(arr, start, mid, end) -> int:
    leftPosition = start
    rightPosition = mid+1

    mergedArray = []
    leftIsBiggerCnt = 0

    while leftPosition <= mid or rightPosition <= end:
        if leftPosition > mid:
            mergedArray.append(arr[rightPosition])
            rightPosition += 1
        elif rightPosition > end:
            mergedArray.append(arr[leftPosition])
            leftPosition += 1
        elif arr[leftPosition] <= arr[rightPosition]:
            mergedArray.append(arr[leftPosition])
            leftPosition += 1
        else:
            mergedArray.append(arr[rightPosition])
            rightPosition += 1
            # left half is already sorted, therefore, if the current number from the left half
            # if bigger the right one, every further number until the mid is also bigger then the
            # current right half's number
            leftIsBiggerCnt += mid - leftPosition + 1 

    for i in range(start, end + 1):
        arr[i] = mergedArray[i - start]

    return leftIsBiggerCnt

def split(arr, start, end) -> int:
    if start == end:
        return 0

    leftIsBiggerCnt = 0

    mid = int((start + end) / 2)
    leftIsBiggerCnt += split(arr, start, mid)
    leftIsBiggerCnt += split(arr, mid + 1, end)

    leftIsBiggerCnt += merge(arr, start, mid, end)
    return leftIsBiggerCnt

def mergeSort(arr):
    return split(arr, 0, len(arr) - 1)

arraySize = readInt()
arr = readIntArray()
print(mergeSort(arr))